Object merge with special characters

rin-st opened the pull request.

rin-st commented on templates/utils.js:
```diff
@@ -26,4 +26,10 @@ export const withDefaults =
+  const str = inspect(val, { depth: null, compact: true, maxArrayLength: null, maxStringLength: null });
+  return str
+    .replace(/"\$\$\$([^"]+)"/g, '$1')
+    .replace(/'\$\$\$([^']+)'/g, '$1')
```
these two lines replaces $$$var with var, so if we meet {a: "$$$var"} or {a: '$$$var'} in code it results {a: var}. Also works for {a: '$$$process.env.B = "true"'} => {a: process.env.B = "true"}

rin-st commented on templates/utils.js:
```diff
@@ -26,4 +26,10 @@ export const withDefaults =
+  return str
+    .replace(/"\$\$\$([^"]+)"/g, '$1')
+    .replace(/'\$\$\$([^']+)'/g, '$1')
+    .replace(/(['"])(.*?\$\{.*?\}.*?)\1/g, '`$2`');
```
if we meet in resulting code "something ${a}" it changes `"` to backtick, so it will be
```
`something ${a}`
```

technophile-04 replied:
maybe let's remove this third thing ` .replace(/(['"])(.*?\$\{.*?\}.*?)\1/g, '`$2`');`? 

If people want to use string interpolation let's suggest them to do it this way: 

https://github.com/technophile-04/hardhat-config/blob/2e7aedcd5fdd03ac7f20f7986d147a6cbe229e4f/extension/packages/hardhat/hardhat.config.ts.args.mjs#L16

rin-st replied:
Umm how it differs for example from `url` fields in this PR [example](https://github.com/scaffold-eth/create-eth/pull/175/files#diff-6df0d91602fb3f0647b7a5dfc9c451fbb3d92c6703ab6569b583811206c7b361R46) ?

The problem is that after merging objects we will get that string with regular `"` or `'` instead of backticks, so we need to convert it. Or am I wrong?

technophile-04 replied:
Ohh right yup missed that part

technophile-04 reviewed (commented):
Love this! and +1 for this solution 🙌 Thanks @rin-st! 

Very small down side of this is extension developers creating extension needs to follow our custom instruction for using variables: 
1. While accessing vars from code snipped they need to use `$$$`
2. If they want to use string interpolation then they need to use escape that var in string `\$`

I think it's probably fine because I feel complex templating engine have it too like handlebars have `{{var}}`  kind of thing?  Also I think most of the file won't require accessing variables from code snippet too, and if some requires then they can use this format. We can document this nicely 🙌. Also with object-merge util their wouldn't be lot of key values where they need to use this format. 

Let's see what others have to say

### To test: 

1. switch to this branch `object-merge-special-characters`

3. run this: 
```shell
yarn build && yarn cli ../test-blah -s hardhat -e technophile-04/hardhat-config:v2 --skip
```

damianmarti commented:
@rin-st this is great and working great too!! This is a simple solution!

Maybe instead of using $$$var for the variables, it is better to use something like $$var$$

rin-st commented:
> Maybe instead of using $$$var for the variables, it is better to use something like $$var$$

Fixed message a bit. Maybe instead of using `$$$var` for the variables, it is better to use something like `$$var$$`. 

---

I'm thinking about just `$$var`. `$$$var` looks too long and regarding `$$var$$` not sure if we need to set the end of expression like that

damianmarti commented:
> > Maybe instead of using $$$var for the variables, it is better to use something like 
> >   v
> >   a
> >   r
> 
> Fixed message a bit. Maybe instead of using `$$$var` for the variables, it is better to use something like `$$var$$`.
> 
> I'm thinking about just `$$var`. `$$$var` looks too long and regarding `$$var$$` not sure if we need to set the end of expression like that

Maybe the end is not needed but I think It's more clear `$$var$$`

technophile-04 commented on templates/solidity-frameworks/hardhat/packages/hardhat/hardhat.config.ts.template.mjs:
```diff
@@ -1,190 +1,207 @@
-import "hardhat-gas-reporter";
-import "solidity-coverage";
-import "@nomicfoundation/hardhat-verify";
-import "hardhat-deploy";
-import "hardhat-deploy-ethers";
-import { task } from "hardhat/config";
-import generateTsAbis from "./scripts/generateTsAbis";
-${imports.filter(Boolean).join("\n")}
-
-// If not set, it uses ours Alchemy's default API key.
-// You can get your own at https://dashboard.alchemyapi.io
-const providerApiKey = process.env.ALCHEMY_API_KEY || "oKxs-03sij-U_N0iOlrSsZFr29-IqbuF";
-// If not set, it uses the hardhat account 0 private key.
-// You can generate a random account with \`yarn generate\` or \`yarn account:import\` to import your existing PK
-const deployerPrivateKey =
-  process.env.__RUNTIME_DEPLOYER_PRIVATE_KEY ?? "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
-// If not set, it uses our block explorers default API keys.
-const etherscanApiKey = process.env.ETHERSCAN_MAINNET_API_KEY || "DNXJA8RX2Q3VZ4URQIWP7Z68CJXQZSC6AW";
-const etherscanOptimisticApiKey = process.env.ETHERSCAN_OPTIMISTIC_API_KEY || "RM62RDISS1RH448ZY379NX625ASG1N633R";
-const basescanApiKey = process.env.BASESCAN_API_KEY || "ZZZEIPMT1MNJ8526VV2Y744CA7TNZR64G6";
+  return deepMerge(target, ...sources);
+}
```
Maybe we should move this to `utils.js`, 

Shall we use some library? Example: https://github.com/fastify/deepmerge ?

technophile-04 commented on templates/solidity-frameworks/hardhat/packages/hardhat/hardhat.config.ts.template.mjs:
```diff
@@ -1,190 +1,207 @@
-        enabled: true,
-        // https://docs.soliditylang.org/en/latest/using-the-compiler.html#optimizer-options
-        runs: 200,
-      },
-    },
-  },
-]`;
+  if (source && typeof source === 'object') {
+    for (const key in source) {
+      if (source[key] && typeof source[key] === 'object') {
+        if (!target[key]) {
+          target[key] = Array.isArray(source[key]) ? [] : {};
+        }
+        deepMerge(target[key], source[key]);
+      } else {
+        target[key] = source[key];
+      }
+    }
+  }
```
Maybe we should move this to `utils.js`, 

Shall we use some library? Example: https://github.com/fastify/deepmerge ?

rin-st replied:
Thanks for the suggestion! See [1762e6b](https://github.com/scaffold-eth/create-eth/pull/175/commits/1762e6ba738006505ae3cedfdbdf534ea4a130df)

technophile-04 reviewed (commented):
> I'm thinking about just `$$var`. `$$$var` looks too long and regarding `$$var$$` not sure if we need to set the end of expression like that

>  Maybe the end is not needed but I think It's more clear `$$var$$`

For me both makes sense so not string opinion. one advantage of `$$var$$` is that it's very hard to clash with user ended code. But even `$$$var` is very hard to clash. 

Also cc @carletex on his views

carletex commented:
No strong feelings either. 

I like `$$var$$` slightly better (even if it's not needed) because it delimits the variable a bit better and it's easier to parse for me

```ts
 mainnet: {
      url: `https://eth-mainnet.alchemyapi.io/v2/\${providerApiKey}`,
      accounts: ["$$$deployerPrivateKey"],
    },
    sepolia: {
      url: `https://eth-sepolia.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
```

But happy either way!

rin-st commented:
> Maybe instead of using `$$$var` for the variables, it is better to use something like `$$var$$`.

> For me both makes sense so not string opinion. one advantage of `$$var$$` is that it's very hard to clash with user ended code. But even `$$$var` is very hard to clash.

> I like `$$var$$` slightly better (even if it's not needed) because it delimits the variable a bit better and it's easier to parse for me

Thanks for the feedback! Changed `$$$var` to `$$var$$`. If you're testing the change with your extension don't forget to update it's variables too

rin-st commented:
Updated `TEMPLATING.md` file so now recommended way is to use `deepMerge`. So
- now merging is a bit easier to implement and read
- we let people now that we have `deepMerge` function ready to use

If you want to test it, use these files (have both `$$var$$` and variables from updated TEMPLATING.md table lines)

`hardhat.config.ts.template.mjs`

<details><summary>Details</summary>
<p>

```
import { withDefaults, stringify, deepMerge } from "../../../../utils.js";

const defaultConfig = {
  solidity: {
    compilers: [
      {
        version: "0.8.20",
        settings: {
          optimizer: {
            enabled: true,
            runs: 200,
          },
        },
      },
    ]
  },
  defaultNetwork: "localhost",
  namedAccounts: {
    deployer: {
      default: 0,
    },
  },
  networks: {
    hardhat: {
      forking: {
        url: `https://eth-mainnet.alchemyapi.io/v2/\${providerApiKey}`,
        enabled: '$$process.env.MAINNET_FORKING_ENABLED === "true"$$',
      },
    },
    mainnet: {
      url: `https://eth-mainnet.alchemyapi.io/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    sepolia: {
      url: `https://eth-sepolia.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    arbitrum: {
      url: `https://arb-mainnet.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    arbitrumSepolia: {
      url: `https://arb-sepolia.g.alchemy.com/v2/\${providerApiKey}`,
        accounts: ["$$deployerPrivateKey$$"],
    },
    optimism: {
      url: `https://opt-mainnet.g.alchemy.com/v2/\${providerApiKey}`,
        accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api-optimistic.etherscan.io",
          apiKey: "$$etherscanOptimisticApiKey$$",
        },
      },
    },
    optimismSepolia: {
      url: `https://opt-sepolia.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api-sepolia-optimistic.etherscan.io",
          apiKey: "$$etherscanOptimisticApiKey$$",
        },
      },
    },
    polygon: {
      url: `https://polygon-mainnet.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    polygonMumbai: {
      url: `https://polygon-mumbai.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    polygonZkEvm: {
      url: `https://polygonzkevm-mainnet.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    polygonZkEvmTestnet: {
      url: `https://polygonzkevm-testnet.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    gnosis: {
      url: "https://rpc.gnosischain.com",
      accounts: ["$$deployerPrivateKey$$"],
    },
    chiado: {
      url: "https://rpc.chiadochain.net",
      accounts: ["$$deployerPrivateKey$$"],
    },
    base: {
      url: "https://mainnet.base.org",
      accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api.basescan.org",
          apiKey: "$$basescanApiKey$$",
        },
      },
    },
    baseSepolia: {
      url: "https://sepolia.base.org",
      accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api-sepolia.basescan.org",
          apiKey: "$$basescanApiKey$$",
        },
      },
    },
    scrollSepolia: {
      url: "https://sepolia-rpc.scroll.io",
      accounts: ["$$deployerPrivateKey$$"],
    },
    scroll: {
      url: "https://rpc.scroll.io",
      accounts: ["$$deployerPrivateKey$$"],
    },
    pgn: {
      url: "https://rpc.publicgoods.network",
      accounts: ["$$deployerPrivateKey$$"],
    },
    pgnTestnet: {
      url: "https://sepolia.publicgoods.network",
      accounts: ["$$deployerPrivateKey$$"],
    },
    celo: {
      url: "https://forno.celo.org",
      accounts: ["$$deployerPrivateKey$$"],
    },
    celoAlfajores: {
      url: "https://alfajores-forno.celo-testnet.org",
      accounts: ["$$deployerPrivateKey$$"],
    },
  },
  etherscan: {
    apiKey: "$$etherscanApiKey$$",
  },
  verify: {
    etherscan: {
      apiKey: "$$etherscanApiKey$$",
    },
  },
  sourcify: {
    enabled: false,
  },
};

const contents = ({ preConfigContent, configOverrides, arrayToSpread, objToMerge }) => {
  // Merge the default config with any overrides
  const finalConfig = deepMerge(defaultConfig, configOverrides[0] || {});
  
  return `import * as dotenv from "dotenv";
dotenv.config();
import { HardhatUserConfig } from "hardhat/config";
import "@nomicfoundation/hardhat-ethers";
import "@nomicfoundation/hardhat-chai-matchers";
import "@typechain/hardhat";
import "hardhat-gas-reporter";
import "solidity-coverage";
import "@nomicfoundation/hardhat-verify";
import "hardhat-deploy";
import "hardhat-deploy-ethers";
import { task } from "hardhat/config";
import generateTsAbis from "./scripts/generateTsAbis";
${preConfigContent[0] || ''}

// If not set, it uses ours Alchemy's default API key.
// You can get your own at https://dashboard.alchemyapi.io
const providerApiKey = process.env.ALCHEMY_API_KEY || "oKxs-03sij-U_N0iOlrSsZFr29-IqbuF";
// If not set, it uses the hardhat account 0 private key.
// You can generate a random account with \`yarn generate\` or \`yarn account:import\` to import your existing PK
const deployerPrivateKey =
  process.env.__RUNTIME_DEPLOYER_PRIVATE_KEY ?? "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
// If not set, it uses our block explorers default API keys.
const etherscanApiKey = process.env.ETHERSCAN_MAINNET_API_KEY || "DNXJA8RX2Q3VZ4URQIWP7Z68CJXQZSC6AW";
const etherscanOptimisticApiKey = process.env.ETHERSCAN_OPTIMISTIC_API_KEY || "RM62RDISS1RH448ZY379NX625ASG1N633R";
const basescanApiKey = process.env.BASESCAN_API_KEY || "ZZZEIPMT1MNJ8526VV2Y744CA7TNZR64G6";

const config: HardhatUserConfig = ${stringify(finalConfig)};

const arrWithAdditionalItems = ${stringify(deepMerge(['a', 'b'], arrayToSpread[0]))};
const mergedObj = ${stringify(deepMerge({ key1: "value1", key2: "value2"}, objToMerge[0]))};

// Extend the deploy task
task("deploy").setAction(async (args, hre, runSuper) => {
  // Run the original deploy task
  await runSuper(args);
  // Force run the generateTsAbis script
  await generateTsAbis(hre);
});

export default config;`;
};

export default withDefaults(contents, {
  preConfigContent: "",
  configOverrides: {},
  arrayToSpread: [],
  objToMerge: {},
});
```

</p>
</details> 

`hardhat.config.ts.args.mjs`

<details><summary>Details</summary>
<p>

```
export const preConfigContent = `
// Custom variables
const CUSTOM_API_KEY = process.env.CUSTOM_API_KEY;
`;

export const configOverrides = {
  networks: {
    hardhat: {
      forking: {
        blockNumber: 1234567
      }
    },
    customNetwork: {
      url: "https://custom.network",
      accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api.custom-explorer.io",
          apiKey: "$$etherscanApiKey$$",
        }
      }
    }
  },
};


export const arrayToSpread = ["Spread", "This"];
export const objToMerge = { key3: "Merged", key4: "Object" };
```

</p>
</details> 

Resulting variables should have these values:

```
const arrWithAdditionalItems = [ 'a', 'b', 'Spread', 'This' ];
const mergedObj = { key1: 'value1', key2: 'value2', key3: 'Merged', key4: 'Object' };
```

technophile-04 commented on contributors/TEMPLATING.md:
```diff
@@ -180,19 +180,19 @@ The special files and folders are:
+| ----------------------- | ----------------------------------------------------------------------------------------------- | ---------------------------------------------------------- | --------------------------------------------------------------------------------------- |
+| Replace an object       | `const replacedObj = ${stringify(replacedObj[0])}`                                              | `const replacedObj = { key1: "Replaced", key2: "Object" }` | `const replacedObj = { key1: "Replaced", key2: "Object" }`                              |
+| Replace an array        | `const replacedArr = ${stringify(replacedArr[0])}`                                              | `const replacedArr = ["Replaced", "Array"]`                | `const replacedArr = ["Replaced", "Array"]`                                             |
+| Object, add new entries | `const mergedObj = ${stringify(deepMerge({ key1: "value1", key2: "value2"}, objToMerge[0] })};` | `const objToMerge = { key3: "Merged", key4: "Object" }`    | `const mergedObj = { key1: "value1", key2: "value2", key3: "Merged", key4: "Object" };` |
```
Shall we mention about deepMerge a bit like how it works? 

<details>

<summary>
    Rules how deepMerge function works:
</summary>

## 1. Primitive Values
When a key exists in both objects with primitive values (strings, numbers, booleans), the value from the second object replaces the first:

```javascript
const deepmerge = createDeepMerge();
const result = deepmerge({a: 1}, {a: 2});
console.log(result); // {a: 2}
```

## 2. Nested Objects
When both objects have a nested object for the same key, it recursively merges them instead of replacing. This preserves properties from both objects:

```javascript
const result = deepmerge(
  {
    config: {
      port: 3000,
      host: 'localhost'
    }
  }, 
  {
    config: {
      port: 8080
    }
  }
);

console.log(result);
// {
//   config: {
//     port: 8080,
//     host: 'localhost'
//   }
// }
```

## 3. New Keys
If a key exists only in the second object, it's added to the result. This ensures all unique properties are preserved:

```javascript
const result = deepmerge({a: 1}, {b: 2});
console.log(result); // {a: 1, b: 2}
```

## 4. Arrays
By default, arrays are concatenated, combining all elements from both arrays in order:

```javascript
const result = deepmerge(
  {list: [1, 2]}, 
  {list: [3, 4]}
);
console.log(result); // {list: [1, 2, 3, 4]}
```

</details>

---- 

Also just thinking out loud there might be the use case where I want to replace the array in deepMerge in nested object instead of concatinate. For example in accounts of `hardhat.config.args.mjs` : 

```js
export const preConfigContent = `
// Custom variables
const CUSTOM_API_KEY = process.env.CUSTOM_API_KEY;
const CUSTOM_ACCOUNT = process.env.CUSTOM_ACCOUNT;
`;

export const configOverrides = {
  networks: {
    optimismSepolia: {
      accounts: ["$$CUSTOM_ACCOUNT$$"],
    },
  },
};

```


I expect final result to be: 

```js
const config: HardhatUserConfig = { 
  networks: {
//    ...otherNetworks, 
    optimismSepolia: {
      accounts: [CUSTOM_ACCOUNT]
    },
  }
}

``` 

But in final result its: `accounts: [deployerPrivateKey, CUSTOM_ACCOUNT]`

Maybe we export another `deepMerge` as `deepMergeArrayReplace` ? 

one of the example from https://github.com/fastify/deepmerge?tab=readme-ov-file#mergearray ?

rin-st replied:
> Shall we mention about deepMerge a bit like how it works?

No strong opinion about it. As I understand it's a well known concept so almost every dev knows what is it and how it works. Checkpoints 1, 2, 3 are default behaviors but we need 100% mention point 4 about merging arrays 

> Also just thinking out loud there might be the use case where I want to replace the array in deepMerge in nested object instead of concatinate. For example in accounts of hardhat.config.args.mjs

We need to be careful about array merging behaviours. In case user wants to replace nested object in
```
const config: HardhatUserConfig = { 
  networks: {
//    ...otherNetworks, 
    optimismSepolia: {
      accounts: [CUSTOM_ACCOUNT]
    },
  }
}
```
merging behavior for other parts will also change, for example for `config.solidity.compilers`.

---

Hmm, for me, overriding full array feels better default behavior for us and also more intuitive for JS devs. For now as I understand we (almost) don't need concatinating behavior for arrays in our extensions. And our second parameter named `configOverrides` so it means it should override everything if possible. 

When override is default, In case someone needs concatenating behavior, like `{ something: [a] } -> { something: [a, b] }`, he can always use `[a, b]` in his args file or just use default deepMerge/change [DeepMergeFn](https://github.com/fastify/deepmerge?tab=readme-ov-file#mergearray).

So, what do you think about making override behavior default?

damianmarti replied:
Yes, I think that making override as default for array is better than merge. 

And it's easy to remove elements from an array in this way.

technophile-04 replied:
> So, what do you think about making override default behavior to override arrays?

> And it's easy to remove elements from an array in this way.

++ yup make more sense + intuitive

rin-st replied:
changed! [7ba7d40](https://github.com/scaffold-eth/create-eth/pull/175/commits/7ba7d40c0e7ec5bac75fed7d74ccb5928952fecf)

technophile-04 reviewed (approved):
Tysm @rin-st!!! This is working great! 

Just updated the example extension to try it out use this: 

```
yarn build:dev && yarn cli -e https://github.com/technophile-04/hardhat-config
```

rin-st commented:
Changed array merge to replace and added example to templating.

To test:
`hardhat.config.ts.template.mjs`

<details><summary>Details</summary>
<p>

```
import { withDefaults, stringify, deepMerge } from "../../../../utils.js";

const defaultConfig = {
  solidity: {
    compilers: [
      {
        version: "0.8.20",
        settings: {
          optimizer: {
            enabled: true,
            runs: 200,
          },
        },
      },
    ]
  },
  defaultNetwork: "localhost",
  namedAccounts: {
    deployer: {
      default: 0,
    },
  },
  networks: {
    hardhat: {
      forking: {
        url: `https://eth-mainnet.alchemyapi.io/v2/\${providerApiKey}`,
        enabled: '$$process.env.MAINNET_FORKING_ENABLED === "true"$$',
      },
    },
    mainnet: {
      url: `https://eth-mainnet.alchemyapi.io/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    sepolia: {
      url: `https://eth-sepolia.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    arbitrum: {
      url: `https://arb-mainnet.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    arbitrumSepolia: {
      url: `https://arb-sepolia.g.alchemy.com/v2/\${providerApiKey}`,
        accounts: ["$$deployerPrivateKey$$"],
    },
    optimism: {
      url: `https://opt-mainnet.g.alchemy.com/v2/\${providerApiKey}`,
        accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api-optimistic.etherscan.io",
          apiKey: "$$etherscanOptimisticApiKey$$",
        },
      },
    },
    optimismSepolia: {
      url: `https://opt-sepolia.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api-sepolia-optimistic.etherscan.io",
          apiKey: "$$etherscanOptimisticApiKey$$",
        },
      },
    },
    polygon: {
      url: `https://polygon-mainnet.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    polygonMumbai: {
      url: `https://polygon-mumbai.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    polygonZkEvm: {
      url: `https://polygonzkevm-mainnet.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    polygonZkEvmTestnet: {
      url: `https://polygonzkevm-testnet.g.alchemy.com/v2/\${providerApiKey}`,
      accounts: ["$$deployerPrivateKey$$"],
    },
    gnosis: {
      url: "https://rpc.gnosischain.com",
      accounts: ["$$deployerPrivateKey$$"],
    },
    chiado: {
      url: "https://rpc.chiadochain.net",
      accounts: ["$$deployerPrivateKey$$"],
    },
    base: {
      url: "https://mainnet.base.org",
      accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api.basescan.org",
          apiKey: "$$basescanApiKey$$",
        },
      },
    },
    baseSepolia: {
      url: "https://sepolia.base.org",
      accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api-sepolia.basescan.org",
          apiKey: "$$basescanApiKey$$",
        },
      },
    },
    scrollSepolia: {
      url: "https://sepolia-rpc.scroll.io",
      accounts: ["$$deployerPrivateKey$$"],
    },
    scroll: {
      url: "https://rpc.scroll.io",
      accounts: ["$$deployerPrivateKey$$"],
    },
    pgn: {
      url: "https://rpc.publicgoods.network",
      accounts: ["$$deployerPrivateKey$$"],
    },
    pgnTestnet: {
      url: "https://sepolia.publicgoods.network",
      accounts: ["$$deployerPrivateKey$$"],
    },
    celo: {
      url: "https://forno.celo.org",
      accounts: ["$$deployerPrivateKey$$"],
    },
    celoAlfajores: {
      url: "https://alfajores-forno.celo-testnet.org",
      accounts: ["$$deployerPrivateKey$$"],
    },
  },
  etherscan: {
    apiKey: "$$etherscanApiKey$$",
  },
  verify: {
    etherscan: {
      apiKey: "$$etherscanApiKey$$",
    },
  },
  sourcify: {
    enabled: false,
  },
};

const contents = ({ preConfigContent, configOverrides, objToDeepMerge, arrayToSpread, objToMerge }) => {
  // Merge the default config with any overrides
  const finalConfig = deepMerge(defaultConfig, configOverrides[0] || {});
  
  return `import * as dotenv from "dotenv";
dotenv.config();
import { HardhatUserConfig } from "hardhat/config";
import "@nomicfoundation/hardhat-ethers";
import "@nomicfoundation/hardhat-chai-matchers";
import "@typechain/hardhat";
import "hardhat-gas-reporter";
import "solidity-coverage";
import "@nomicfoundation/hardhat-verify";
import "hardhat-deploy";
import "hardhat-deploy-ethers";
import { task } from "hardhat/config";
import generateTsAbis from "./scripts/generateTsAbis";
${preConfigContent[0] || ''}

// If not set, it uses ours Alchemy's default API key.
// You can get your own at https://dashboard.alchemyapi.io
const providerApiKey = process.env.ALCHEMY_API_KEY || "oKxs-03sij-U_N0iOlrSsZFr29-IqbuF";
// If not set, it uses the hardhat account 0 private key.
// You can generate a random account with \`yarn generate\` or \`yarn account:import\` to import your existing PK
const deployerPrivateKey =
  process.env.__RUNTIME_DEPLOYER_PRIVATE_KEY ?? "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
// If not set, it uses our block explorers default API keys.
const etherscanApiKey = process.env.ETHERSCAN_MAINNET_API_KEY || "DNXJA8RX2Q3VZ4URQIWP7Z68CJXQZSC6AW";
const etherscanOptimisticApiKey = process.env.ETHERSCAN_OPTIMISTIC_API_KEY || "RM62RDISS1RH448ZY379NX625ASG1N633R";
const basescanApiKey = process.env.BASESCAN_API_KEY || "ZZZEIPMT1MNJ8526VV2Y744CA7TNZR64G6";

const config: HardhatUserConfig = ${stringify(finalConfig)};

const arrWithAdditionalItems = ${stringify(['a', 'b', ...arrayToSpread[0]])};
const mergedObj = ${stringify(deepMerge({ key1: "value1", key2: "value2"}, objToMerge[0]))};
const deepMergedObj = ${stringify(deepMerge({ key1: "value1", key2: "value2", key3: ["value3_0", "value3_1"]}, objToDeepMerge[0] ))};

// Extend the deploy task
task("deploy").setAction(async (args, hre, runSuper) => {
  // Run the original deploy task
  await runSuper(args);
  // Force run the generateTsAbis script
  await generateTsAbis(hre);
});

export default config;`;
};

export default withDefaults(contents, {
  preConfigContent: "",
  configOverrides: {},
  objToDeepMerge: {},
  arrayToSpread: [],
  objToMerge: {},
});
```

</p>
</details> 


`hardhat.config.ts.template.mjs`

<details><summary>Details</summary>
<p>


```
export const preConfigContent = `
// Custom variables
const CUSTOM_API_KEY = process.env.CUSTOM_API_KEY;
`;

export const configOverrides = {
  solidity: {
    compilers: [
      {
        version: "0.8.21",
        settings: {
          optimizer: {
            enabled: true,
          }
        }
      }
    ]
  },
  networks: {
    hardhat: {
      forking: {
        blockNumber: 1234567
      }
    },
    customNetwork: {
      url: "https://custom.network",
      accounts: ["$$deployerPrivateKey$$"],
      verify: {
        etherscan: {
          apiUrl: "https://api.custom-explorer.io",
          apiKey: "$$etherscanApiKey$$",
        }
      }
    }
  },
};

export const arrayToSpread = ["Spread", "This"];
export const objToMerge = { key3: "Merged", key4: "Object" };
export const objToDeepMerge = { key1: "replacedValue_1", key3: ["replacedValue3_0"], key4: "value4" };
```

</p>
</details> 


Resulting variables should have these values:

```
const arrWithAdditionalItems = ["a", "b", "Spread", "This"];
const mergedObj = { key1: "value1", key2: "value2", key3: "Merged", key4: "Object" };
const deepMergedObj = { key1: "replacedValue_1", key2: "value2", key3: ["replacedValue3_0"], key4: "value4" };
```

technophile-04 commented:
So added a breaking change changeset to this at https://github.com/scaffold-eth/create-eth/pull/175/commits/5c730d6d3e8a4d5877b4feb9d235dac7c1c5a4df and changed base branch to [beta](https://github.com/scaffold-eth/create-eth/tree/beta) branch. 

### The workflow until we officially launch the major version: 

> major version will be released when we merge `beta` branch into `main` branch. 

So the `main` and `beta` branches will run parallel. We merge `patch`, `minor` versions to `main` things like backmerge Pr's from SE-2 and keep releasing them as patch, minor versions (0.x.x). 

In `beta` branch we merge stuff like this. Once anything is merged in `beta` branch it will release the beta version to npm `1.x.x-beta.x`. 

We also make sure we keep backmerging `main` -> `beta` so that's it not left behind and we don't face lot of merge conflicts while merging `beta` -> `main`

![image](https://github.com/user-attachments/assets/e7a790b8-d7be-448a-acea-c677d25fbcba)

--- 

I played around configuring and tinkering this workflow over weekend in https://github.com/technophile-04/test-changeset-flow-package/ and setup similar things in `beta` here in this repo. 

1. merging breaking change to beta branch https://github.com/technophile-04/test-changeset-flow-package/pull/15
2. This is how "Version packages" PR look for beta branch : https://github.com/technophile-04/test-changeset-flow-package/pull/16
3. Backmerging `main` to `beta` branch at https://github.com/technophile-04/test-changeset-flow-package/pull/17
...
4. once we are ready to launch major version we create `beta` -> `main` [PR](https://github.com/technophile-04/test-changeset-flow-package/pull/18)  and when ready we exit beta mode by doing `yarn chageset pre exit`.
Merge the PR. 
5. After merging github actions will create "Version packages" for `main` branch which will contain major version release
6. After merging above PR it will release a new version. 

--- 

Would love to know your thoughts / suggestion, also happy to take any other approaches as well 🙌 

We could do same thing for #56 as well 🙌

rin-st commented:
Thank you a lot for the scheme and explanation! Looking very good!

Some questions
> In beta branch we merge stuff like this. Once anything is merged in beta branch it will release the beta version to npm 1.x.x-beta.x.

Is releasing new beta version working for backmerges from `main`?

What 1.x.x exactly? Where it takes version? Shouldn't it be 1.0.0-beta.x?

Can we  use  `beta` branch for minor/patch versions?

technophile-04 commented:
> What 1.x.x exactly? Where it takes version? Shouldn't it be 1.0.0-beta.x?

Yup it seems any changes merged to beta branch will increase 1.0.0-beta.x the 'x' value, so it will be 1.0.0-beta.1, 1.0.0-beta.2 etc 


> Can we use beta branch for minor/patch versions?

Ohh you mean to `beta` branch workflow for releasing minor/patch version of `main` branch we well so that we release beta patch/minor version for testing and then merge it to main finally? Yes we can actually maybe we call it release-branch? that's what #56 we were planning to conquer....lol I hope I got your question correctly or you mean to say releasing the minor/patch version for beta branch? Merging anything to beta branch will only increment 1.0.0-beta.x so won't change minor/patch version their, since it makes sense? Because we want to accumulate all changes and release as one big version? 


> Is releasing new beta version working for backmerges from main?

Ohh yeah when making backmerges PR from `main` -> `beta` if we add changeset to that PR it will release a new beta version as 1.0.0-beta.x. 

note while making backmerge we should make sure we have already merged "Version packages" PR to `main` which cleans up `.changesets` dir in `main` branch and then create backmerge PR to `beta`

carletex commented:
I think this is great!! Love the graphic <3

The workflow makes a lot of sense. Let's try to not extend this too much in time, since there is a lot of time consuming back-merging here haha. But this is the safest way!

So let's see if I get TLDR correctly:
- Let's keep going as usual (SE-2 back merges, cli changes, etc)
- Let's run a parallel BETA branch that will publish the beta version (make sure we back-merge main changes). Let's add all the breaking changes to this branch.
- When ready:
  - Let's prepare the extension to work with the breaking changes (e.g. in a __beta branch) + test with the beta deployed version
  - Merge beta => main.

Am I missing something?

technophile-04 commented:
> Let's try to not extend this too much in time, since there is a lot of time consuming back-merging here haha. 

yeah felt that too and love the TLDR;! Yes that's just precise and exactly the plan 🙌 Thanks!
